### DISCUSSION TITLE
Deal with 2-D heat transfer problem, can't get the ideal result

### DISCUSSION THREAD

**Discussion by w-venton** [lululxvi/deepxde#1236] [Q&A]
Hi everyone.
I am using deepxde to solve a two dimensional heat transfer problem, the following diagram shows the settings for some **physical cases**, I have made some attempts but the results are not very good. 
![7b8a9b63ce44c284aa43031a8737dfb](https://user-images.githubusercontent.com/84429872/230084805-a62b74d7-c4d7-4022-9a5c-001a97f6d152.jpg)

1. **how to set these values for a heat transfer problem at the μm level**, I have not found any documentation in deepxde that relates to units; 
2. if it is possible to **write the code like this for a case with three materials**, and if there are any problems, please help me to point them out; 
3. how to **adapt the power of a heat source to my code when adding a After adding Q**, my loss has become very large. Do you all have any good suggestions?

How can I rescale, should I rescale the size? what's the meaning of making it [0,1]x[0,1]? I‘m little confused. Or should I rescale the T, just like T 30K, then output transform like lambda x,y: y*10? 

I would really, really like to have your help!!

My code:
```python
import deepxde as dde
import matplotlib.pyplot as plt
import numpy as np
import matplotlib.animation as animation
from matplotlib.animation import FuncAnimation
import tensorflow as tf
import pandas as pd

DoCompare=False
Hmetal = 0.45  # unit is μm ??
Wmetal = 1
h = 5
wsio = 10
hsio = 8
hsub = 2
x_point = (wsio - Wmetal) / 2  
Wtotal = wsio 
Htotal = hsub+hsio 

t1 = 0
t2 = 4
myT0 = 30  
Ti = 30  
end_time = 4
# h   (2e13 0); // W/m^3 == kg/m/s^3 

# Thermal diffusivity of silicon
a1 = 380 / (8940 * 385)  #copper k/ρc
a2 = 1.4 / (2200 * 966)  # SIO2
a3 = 148 / (2200 * 966)  #sub

Q_heat = 10  #Q_heat:  f(x,y,t)=J^2*ρ/(c*ρ) in area, else is 0


def pde(X, T):
    dT_X = tf.gradients(T, X)[0]
    dT_x = dT_X[:, 0:1]
    dT_y = dT_X[:, 1:2]
    dT_t = dT_X[:, 2:3]
    dT_xx = tf.gradients(dT_x, X)[0][:, 0:1]
    dT_yy = tf.gradients(dT_y, X)[0][:, 1:2]

    f_heat = tf.where((X[:, 0:1] > x_point) & (X[:, 0:1] < x_point + Wmetal) &
                      (X[:, 1:2] > hsub + h)
                      & (X[:, 1:2] < hsub + h + Hmetal),
                      X[:, 0:1] * 0 + Q_heat, X[:, 0:1] * 0 + 0)

    a = tf.where((X[:, 0:1] > x_point) & (X[:, 0:1] < x_point + Wmetal) &
                 (X[:, 1:2] > hsub + h)
                 & (X[:, 1:2] < hsub + h + Hmetal), X[:, 0:1] * 0 + a1,
                 X[:, 0:1] * 0 + a2)  #metal

    a = tf.where((X[:, 1:2] < hsub), X[:, 0:1] * 0 + a3, a)  #sub

    return (dT_t - (a * (dT_xx + dT_yy)) - f_heat)


# bounday
def r_boundary(X, on_boundary):
    x, y, t = X
    return on_boundary and np.isclose(x, 10)


def l_boundary(X, on_boundary):
    x, y, t = X
    return on_boundary and np.isclose(x, 0) 


def up_boundary(X, on_boundary):
    x, y, t = X
    return on_boundary and np.isclose(y, 10)


def down_boundary(X, on_boundary):
    x, y, t = X
    return on_boundary and np.isclose(y, 0)


def boundary_initial(X, on_initial):
    x, y, t = X
    return on_initial and np.isclose(t, 0)


def init_func(X):
    return myT0 * np.ones((len(X), 1))


num_domain = 62500
num_boundary = 8000
num_initial = 20000
#num_test = 62500
layer_size = [3] + [80] * 5 + [1]
activation_func = "tanh"
initializer = "Glorot uniform"
lr = 1e-3

# Applying Loss weights as given below
loss_weights = [1, 1, 1, 1, 1e-2, 1e-2]
iterations = 15000
optimizer = "adam"
batch_size_ = 256

geom = dde.geometry.Rectangle([0, 0], [10, 10])
timedomain = dde.geometry.TimeDomain(0, end_time)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

bc_l = dde.NeumannBC(geomtime, lambda x: 0, l_boundary)  
bc_r = dde.NeumannBC(geomtime, lambda x: 0, r_boundary)
bc_up = dde.NeumannBC(geomtime, lambda x: 0, up_boundary)
bc_low = dde.DirichletBC(geomtime, lambda x: myT0,
                         down_boundary)  #300K

ic = dde.IC(geomtime, init_func, boundary_initial)

bcs = [bc_l, bc_r, bc_up, bc_low, ic]
data = dde.data.TimePDE(geomtime,
                        pde,
                        bcs,
                        num_domain=num_domain,
                        num_boundary=num_boundary,
                        num_initial=num_initial,num_test=10000
                        )

net = dde.nn.tensorflow_compat_v1.fnn.FNN(layer_size, activation_func,
                                          initializer)

net.apply_output_transform(lambda x, y: abs(y))
model = dde.Model(data, net)
model.compile(optimizer, lr=lr, loss_weights=loss_weights)

# To save the best model every 1000 epochs
checker = dde.callbacks.ModelCheckpoint("./2D/model/Qheat.ckpt",
                                        save_better_only=True,
                                        period=1000)
losshistory, trainstate = model.train(
    iterations=iterations,
    batch_size=batch_size_,
    callbacks=[checker],
    #update_every=100,
    display_every=1000,
)
```

I really don't know where the problem is? the **result in TCAD** is below:
![image](https://user-images.githubusercontent.com/84429872/230082967-c2bc7997-28e4-424e-99b2-12e8f8d6e23b.png)

But my result is like below, when `Q_heat = 2e7/ (8940 * 385)`:
![image](https://user-images.githubusercontent.com/84429872/230083202-47073485-5d86-4efa-ad2e-7f153f24caee.png)

when `Q_heat = 2e7/ (8940 * 385)`:
![image](https://user-images.githubusercontent.com/84429872/230087787-d15ad482-c5fd-4fc6-b0c1-1a70362626d4.png)
And the loss is : 
![image](https://user-images.githubusercontent.com/84429872/230087927-651a4ac7-8ed7-4626-b4df-5cf80ad414ca.png)
The loss of PDE didn't change...
I really need your help!!! please...
@lululxvi

**Comment by lululxvi** [comment]
Check FAQ Q: I failed to train the network or get the right solution, e.g., large training loss, unbalanced losses.

**Comment by w-venton** [comment]
I've read all of this FAQ, but I still don't quite understand how to reduce thescale of PDE. For example, divide the whole by a coefficient?
What's more, I don't quite understand how to make the material junction meet the continuous heat flux constraint, if the situation of different media is represented by a region? Or do I need to use two regions to represent two materials, so how should this condition of continuous heat flux be expressed by the constraint in deepxde? Use OperatorBC?

**Comment by lululxvi** [comment]
- https://en.wikipedia.org/wiki/Nondimensionalization
- Different materials: mathematically, it is just a discontinuous function.

**Comment by MohammdReza5106** [comment]
@w-venton 
Did you figure it out? I'm dealing with a somewhat similar issue, trying to work it out. It occurred to me that if we take all the parameters into account and scale them from 1 to 0, it might be helpful.
